### PR TITLE
Refact: Refactored Forced decision

### PR DESCRIPTION
**Pull request by msohailhussain**
## Summary
- Removed `get_variation_from_experiment_rule` and `get_variation_from_delivery_rule` from decision service. 
- `flag_variations_map` logic improved using existing loops and removed `get_all_variations_for_each_rule`.
- Unit test fixed.

## Test plan
All FSC and unit tests should pass

### CODE REVIEW

File: optimizely/project_config.py [hunk 1]
```diff
@@ -100,6 +100,7 @@ def __init__(self, datafile, logger, error_handler):
         self.variation_variable_usage_map = {}
         self.variation_id_map_by_experiment_id = {}
         self.variation_key_map_by_experiment_id = {}
+        self.flag_variations_map = {}
```
Comment (msohailhussain) [inline optimizely/project_config.py:103]: holding this PR for now...

File: optimizely/project_config.py [hunk 2]
```diff
@@ -215,6 +195,28 @@ def get_rollout_experiments(self, rollout):
 
         return rollout_experiments
 
+    def get_rules_for_flag(self, feature_flag):
```
Comment (msohailhussain) [inline optimizely/project_config.py:198]: doesn't make sense to add this method.

File: optimizely/project_config.py [hunk 3]
```diff
@@ -215,6 +195,28 @@ def get_rollout_experiments(self, rollout):
...
+        rules = map(lambda exp_id: self.experiment_id_map[exp_id], feature_flag['experimentIds'])
+        rules = list(rules)
+        rollout = None if len(feature_flag['rolloutId']) == 0 else self.rollout_id_map[feature_flag['rolloutId']]
+
+        if rollout:
+            rollout_experiments = rollout.experiments
+            for exp in rollout_experiments:
+                rules.append(self.experiment_id_map[exp['id']])
```
Comment (jaeopt) [inline optimizely/project_config.py:206]: I understand experiment_id_map is for ab-tests only, not including rollout rules. Did it pass our tests?

File: optimizely/project_config.py [hunk 4]
```diff
@@ -120,46 +121,25 @@ def __init__(self, datafile, logger, error_handler):
                 )
 
         self.feature_key_map = self._generate_key_map(self.feature_flags, 'key', entities.FeatureFlag)
-
+        self.flag_variations_map = self.generate_feature_variation_map(self.feature_flags)
```
Comment (jaeopt) [inline optimizely/project_config.py:124]: Remove line 103

**Review by jaeopt — CHANGES_REQUESTED** [review]
One change suggested

File: optimizely/decision_service.py [hunk 5]
```diff
@@ -517,11 +450,37 @@ def get_variation_for_feature(self, project_config, feature, user_context, optio
...
-                if experiment:
-                    variation, variation_reasons = self.get_variation_from_experiment_rule(
-                        project_config, feature.key, experiment, user_context, options)
-                    decide_reasons += variation_reasons
-                    if variation:
-                        return Decision(experiment, variation, enums.DecisionSources.FEATURE_TEST), decide_reasons
+                decision_variation = None
 
+                if experiment:
+                    # variation, variation_reasons = self.get_variation_from_experiment_rule(
```
Comment (mnoman09) [inline optimizely/decision_service.py:456]: Remove this commented code

File: optimizely/project_config.py [hunk 6]
```diff
@@ -120,46 +121,39 @@ def __init__(self, datafile, logger, error_handler):
...
-                for exp_id in flag['experimentIds']:
-                    experiments.append(self.experiment_id_map[exp_id])
-            if not flag['rolloutId'] == '':
-                rollout = self.rollout_id_map[flag['rolloutId']]
-
-                rollout_experiments = self.get_rollout_experiments(rollout)
-
-                if rollout and rollout.experiments:
-                    experiments.extend(rollout_experiments)
+        # Dict containing map of experiment ID to feature ID.
```
Comment (mnoman09) [inline optimizely/project_config.py:133]:
```suggestion
        # Dictionary containing dictionary of experiment ID to feature ID.
```

File: optimizely/decision_service.py (hunk 5 above)
Comment (The-inside-man) [inline optimizely/decision_service.py:456]: Should we remove commented out code?

File: optimizely/project_config.py [hunk 7]
```diff
@@ -120,46 +121,39 @@ def __init__(self, datafile, logger, error_handler):
...
+        # for checking that experiment is a feature experiment or not.
+        self.experiment_feature_map = {}
+        for feature in self.feature_key_map.values():
```
Comment (mnoman09) [inline optimizely/project_config.py:136]: Why are we running this loop again see line 127 can we reuse it?

**Review by mnoman09 — CHANGES_REQUESTED** [review]
Added some comments Ovarall looks good.

File: optimizely/decision_service.py (hunk 5 above)
Comment (ozayr-zaviar) [reply optimizely/decision_service.py:457]: I was testing this code with travis. Will remove it on finalization.

File: optimizely/project_config.py (hunk 3 above)
Comment (ozayr-zaviar) [reply optimizely/project_config.py]: All unit tests and FSC is passing.

**Review by jaeopt — APPROVED** [review]
LGTM

**Review by The-inside-man — APPROVED** [review]
All lgtm, seems a few tests are failing in CI for pypy3, even after rerunning several times. I dont see this as a concern right now and can be addressed later as this seems to be a common issue with these versions of Python.

**Comment by msohailhussain** [comment]
> All lgtm, seems a few tests are failing in CI for pypy3, even after rerunning several times. I dont see this as a concern right now and can be addressed later as this seems to be a common issue with these versions of Python.

true.

**Pull request merged**